### PR TITLE
Add Insane Comparison Rule

### PR DESCRIPTION
**Pull request by b1rdex**
This is a follow up for https://github.com/phpstan/phpstan/discussions/4359

### CODE REVIEW

File: conf/config.level5.neon [hunk 1]
```diff
@@ -19,3 +19,10 @@ services:
 			reportMaybes: %reportMaybes%
 	-
 		class: PHPStan\Rules\DateTimeInstantiationRule
+	-
+		class: PHPStan\Rules\Comparison\InsaneComparisonRule
+		arguments:
+			treatMixedAsPossibleString: %reportMaybes%
```
Comment (b1rdex) [inline conf/config.level5.neon:25]: I'm not sure whether this should be switchable, I'd prefer always enabling it

File: src/Rules/Comparison/InsaneComparisonRule.php [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+		) {
+			return [];
+		}
+
+		$left_type = $scope->getType($node->left);
+		$right_type = $scope->getType($node->right);
+
+		// if ($left_type instanceof ErrorType || $right_type instanceof ErrorType) {
+		//     return [];
+		// }
```
Comment (b1rdex) [inline src/Rules/Comparison/InsaneComparisonRule.php:54]: Should I uncomment this?

File: tests/PHPStan/Rules/Comparison/data/insane-comparison-maybe.php [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+/**
+ * @param mixed[]  $array
+ */
+function z(array $array, string $string): void {
```
Comment (b1rdex) [inline tests/PHPStan/Rules/Comparison/data/insane-comparison-maybe.php:6]:
```suggestion
function z(array $array): void {
```

**Comment by b1rdex** [comment]
Hi @ondrejmirtes. Could you please take a look at this?

**Comment by ondrejmirtes** [comment]
Hi, this code will need some work but I'm currently busy to look into this. Stay tuned!

**Comment by b1rdex** [comment]
Hi @ondrejmirtes. I'd like to finish the thing and have it merged. Any chance having this reviewed?

**Comment by ondrejmirtes** [comment]
Hi, it'd need some more work to be able to merge this. For example I don't want the new `treatMixedAsPossibleString` setting. Instead, the rule should employ RuleLevelHelper which returns StrictMixedType with `checkExplicitMixed: true`.

RuleLevelHelper would also make sure that some errors are reported on level 5 while others (when partially wrong unions are involved) are reported on level 7.

Additionally, the error message needs improvement by saying what's wrong about the comparison.

**Comment by ondrejmirtes** [comment]
Hi, I'm cleaning up old and stale PRs. Please send a new PR if you're still interested, thanks.

**Comment by b1rdex** [comment]
Hi @ondrejmirtes! I don't plan to reimplement this.

This rule was about https://wiki.php.net/rfc/string_to_number_comparison – breaking change in PHP 8. I think this is a good feature, but it's hard to implement it.

Personally, I think it's easier to just create a rule `don't ever use a loose comparison operators if variables are of different types (or mixed)` (including array_* functions and more stuff – see Proposal in the RFC) and add it to some PHPStan level.

**Pull request closed without merging**